source/Img32.pas

```diff
@@ -659,6 +667,46 @@ procedure NormalizeAngle(var angle: double; tolerance: double = Pi/360);
 end;
 //------------------------------------------------------------------------------
 
+{$IFDEF CPUX86}
+{ Trunc with FPU code is very slow because the x87 ControlWord has to be changed
+  and then there is Delphi's Default8087CW variable that is not thread-safe. }
+
+//__Trunc: An efficient Trunc() algorithm (ie rounds toward zero)
+function __Trunc(Value: Double): Integer;
+type
+  TDoubleDataRec = packed record // x86 -> Little Endian
+    Mantisse: Integer;
+    Exp: Integer;
+  end;
+var
+  exp: integer;
+  i64: UInt64 absolute Value;
+  dbl: TDoubleDataRec absolute Value;
+begin
+  //https://en.wikipedia.org/wiki/Double-precision_floating-point_format
+  Result := 0;
+  if i64 = 0 then Exit;
+  exp := Integer(Cardinal(i64 shr 52) and $7FF) - 1023;
+  //nb: when exp == 1024 then Value == INF or NAN.
+  if exp < 0 then
+    Exit
+  else if exp > 52 then
```

Comment:
A couple of other observations here ...
1. `exp > 31` would create an integer overflow so testing for `exp > 52` seems redundant (unless the result was changed to a 64bit int). Anyhow I'm somewhat inclined to either skip testing for exp > 52 (or exp > 31), as I doubt that would ever be achieved given how this function is used. 
2. It seems that replacing `or (UInt64(1) shl exp)` in line 696 with an integer array lookup is more efficient.

Reply:
That function, except for the last line is the version that was committed and then removed in previous commit. I only revived it.

If "exp" never exceeds 31 then you can just use "Cardinal(1) shl exp" which should be even fast than a lookup.

Reply:
> That function, except for the last line, is the version that was removed in a previous commit. I only revived it.

I'm fully aware that this was my own code 😁. I can only blame you 🤣 for making me look again at stuff I've written 😱 that can improved, if only with better inline documentation.